### PR TITLE
feat(VTimeline): update to v3

### LINKED ISSUE #5360: [Feature Request] add support for horizontal v-timeline
### Problem to solve
If you have a list of object that each requires a timeline, or if your design requires the old fashioned horizontal timeline, in those case the actual vertical timeline don't do the trick
(I was very exited by the Vtimeline. I still didn't had time to really look into it but I'm kind of disappointed that without horizontal layout i can't use it in my project right now)

### Proposed solution
Maybe allow us by a props to switch the layout of the Vtimeline to horizontal

### LINKED ISSUE #6594: [Feature Request] Add alignment options to v-timeline
### Problem to solve

With the current framework options there's (as far as I know) no way to alter the 50%/50% split between the content and the `opposite` slot.

See https://github.com/vuetifyjs/vuetify/issues/6594#issuecomment-468389266

### Proposed solution

I think this needs to be discussed as I don't have a good idea for this.

### LINKED ISSUE #9370: [Bug Report] v-timeline with dense and left attribute not working
### Environment
**Vuetify Version:** 2.1.4
**Vue Version:** 2.6.10
**Browsers:** Chrome 76.0.3809.132
**OS:** Linux

### Steps to reproduce
Make a simple timeline and add dense property then use the left attribute to make timeline items get oriented in only 1 side.

### Expected Behavior
All items to be on the side I define (left or right)

### Actual Behavior
Attributes (left and right) are not working with dense enabled

### Reproduction Link
<a href="https://codepen.io/rapper/pen/wvvWzqg" target="_blank">https://codepen.io/rapper/pen/wvvWzqg</a>

### LINKED ISSUE #11721: [Feature Request] v-timeline: end lines at first and last item
### Problem to solve
Currently, v-timeline always extends the line beyond the first and last dot. I would like the line to end at the dots, so it looks similar to v-stepper. 

### Proposed solution
Add a prop to v-timeline which allows us to choose between an extended line and truncated line.

### LINKED ISSUE #12971: [Bug Report] VTimelineItem Icon always gets dark theme color (white)
### Environment
**Vuetify Version:** 2.4.2
**Vue Version:** 2.6.11
**Browsers:** Chrome 87.0.4280.141
**OS:** Windows 10

### Steps to reproduce
Go to https://vuetifyjs.com/en/components/timelines/#icon-dots
Switch to light or dark mode.(Icon is always white)

### Expected Behavior
Icon color should follow theme. Or is this the expected behaviour?

### Actual Behavior
Set the appriopriate icon color from theme.

### Reproduction Link
<a href="https://codepen.io/bfritscher/pen/vYXvKzx" target="_blank">https://codepen.io/bfritscher/pen/vYXvKzx</a>

### Other comments
In https://github.com/vuetifyjs/vuetify/blob/master/packages/vuetify/src/components/VTimeline/VTimelineItem.ts
Icon is created with this code:
```js
  return this.$createElement(VIcon, {
        props: {
          color: this.iconColor,
          dark: !this.theme.isDark,
          small: this.small,
        },
      }, this.icon)
```

Why is dark inverted? And no light propagated?

### PR DESCRIPTION
**Pull request by nekosaur**
## Description
update v-timeline to v3

fixes #9370 
fixes #5360 
fixes #12971 
fixes #11721 
fixes #6594

## Breaking changes

v-timeline now uses provide/inject to keep track of v-timeline-items. This means that if you are dynamically toggling items, you will need to use `index` prop to re-insert them in the correct position.

removed v-card wedge css 

### v-timeline
- added `direction` prop
- added `tag` prop
- added `truncate-line` prop
- `dense` prop replaced by `side` (and `density`)
- removed `align-top`
- removed `light` and `dark` props
- removed `reverse` prop

### v-timeline-item
- added `hide-opposite`
- added `elevation`
- added `tag`
- removed `light` and `dark` props
- replaced `small` and `large` props with `size`
- removed `left` and `right` props

## Motivation and Context

v3

## How Has This Been Tested?

playground, unit tests

## Markup:

<details>

```vue
<template>
  <v-app class="fill-height" :theme="theme">
    <v-locale-provider :rtl="rtl">
      <div style="display: flex; flex-direction: column">
        <input type="checkbox" :checked="rtl" @change="rtl = !rtl">rtl
        <input type="checkbox" :checked="show" @change="show = !show">show
        <input type="checkbox" :checked="theme === 'dark'" @change="theme = theme === 'dark' ? 'light' : 'dark'">theme
        <input type="checkbox" :checked="!!icon" @change="icon = icon == null ? '$close' : null">icon
        <select v-model="direction" class="input">
          <option value="vertical">vertical</option>
          <option value="horizontal">horizontal</option>
        </select>
        <select v-model="density" class="input">
          <option value="compact">compact</option>
          <option value="comfortable">comfortable</option>
          <option value="default">default</option>
        </select>
        <select v-model="side" class="input">
          <option value="start">start</option>
          <option value="end">end</option>
          <option :value="null">null</option>
        </select>
        <input v-model="thickness" type="number"/>
        <input v-model="inset" type="number"/>
        <select v-model="truncate">
          <option value="none">none</option>
          <option value="both">both</option>
          <option value="start">start</option>
          <option value="end">end</option>
        </select>
      </div>
      <v-container style="height: 100%">
        <v-row
          align="center"
          justify="space-between"
          style="height: 100%"
        >
          <v-timeline :direction="direction" :density="density" :line-thickness="thickness" :side="side" :line-inset="inset" line-color="red-lighten-2" :truncate-line="truncate">
            <v-timeline-item size="x-large" :icon="icon">
              <template v-slot:opposite>
                opposite 1
              </template>
              timeline item 1
            </v-timeline-item>
            <v-timeline-item size="large" :icon="icon">
              <template v-slot:opposite>
                opposite 2
              </template>
              timeline item 2
            </v-timeline-item>
            <v-timeline-item :icon="icon">
              <template v-slot:opposite>
                opposite 3
              </template>
              timeline item 3
            </v-timeline-item>
            <v-timeline-item size="small" :icon="icon" v-if="show" :index="3">
              <template v-slot:opposite>
                opposite 4
              </template>
              timeline item 4
            </v-timeline-item>
            <v-timeline-item size="x-small" :icon="icon">
              <template v-slot:opposite>
                opposite 5
              </template>
              timeline item 5
            </v-timeline-item>
          </v-timeline>
        </v-row>
      </v-container>
    </v-locale-provider>
  </v-app>
</template>

<script>
  import { ref } from 'vue'

  export default {
    name: 'Playground',
    setup () {
      const show = ref(true)
      const reverse = ref(false)
      const icon = ref(null)
      const rtl = ref(false)
      const truncate = ref('both')
      const direction = ref('vertical')
      const density = ref('default')
      const side = ref(null)
      const thickness = ref('10')
      const inset = ref(null)
      const theme = ref('light')

      return { theme, show, reverse, icon, rtl, truncate, direction, density, side, thickness, inset }
    },
  }
</script>

<style scoped>
  .select {
    border: 1px solid black;
    margin: 4px;
  }
</style>

```
</details>

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Improvement/refactoring (non-breaking change that doesn't add any features but makes things better)

## Checklist:

- [x] The PR title is no longer than 64 characters.
- [x] The PR is submitted to the correct branch (`master` for bug fixes and documentation updates, `dev` for new features and backwards compatible changes and `next` for non-backwards compatible changes).
- [x] My code follows the code style of this project.
- [x] I've added relevant changes to the documentation (applies to new features and breaking changes in core library)

### CODE REVIEW

**Review by KenG-pro — APPROVED** [review]

**Comment by johnleider** [comment]
A few ideas:

## Timeline Divider
* inset
image
* thickness
image
* color
image

## Timeline Item
* avatar
image
* icon
image
* offset
image

**Comment by johnleider** [comment]
I think we should use density to control the layout of items:

Default would show both sides by default
image

Comfortable would position all cards to the right and all opposite slots to the left
image

Compact would hide the opposite slot
image

File: packages/vuetify/src/components/VTimeline/VTimeline.tsx [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+        ]}
+        style={{
+          '--v-timeline-line-thickness': convertToUnit(props.lineThickness),
+          '--v-timeline-line-inset': convertToUnit(props.lineInset || undefined),
+        }}
+      >
+        { props.truncateLine === 'none' || props.truncateLine === 'end' ? (
+          <VTimelineItem hideDot />
+        ) : undefined }
+        { ctx.slots.default?.() }
```
Comment (johnleider) [inline packages/vuetify/src/components/VTimeline/VTimeline.tsx:100]:
```suggestion
        { props.truncateLine === 'none' || props.truncateLine === 'end' && (
          <VTimelineItem hideDot />
        ) }

        { slots.default?.() }
```

File: packages/vuetify/src/components/VTimeline/VTimelineDivider.tsx [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+// Utilities
+import { defineComponent, inject } from 'vue'
+
+export default defineComponent({
+  name: 'VTimelineDivider',
+
+  props: {
+    hideDot: Boolean,
+    lineColor: {
```
Comment (johnleider) [inline packages/vuetify/src/components/VTimeline/VTimelineDivider.tsx:19]: remove default

File: packages/vuetify/src/components/VTimeline/VTimelineDivider.tsx [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+      type: String,
+      default: 'secondary',
+    },
+    icon: String,
+    iconColor: String,
+    fillDot: Boolean,
+    dotColor: {
```
Comment (johnleider) [inline packages/vuetify/src/components/VTimeline/VTimelineDivider.tsx:26]: remove default

File: packages/vuetify/src/components/VTimeline/VTimelineDivider.tsx [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+                roundedClasses.value,
+                backgroundColorClasses.value,
+              ]}
+              style={ backgroundColorStyles.value }
+            >
+              {
+                ctx.slots.default ? ctx.slots.default({ icon: props.icon, iconColor: props.iconColor, size: props.size })
+                : props.icon ? <VIcon icon={ props.icon } color={ props.iconColor } size={ props.size } />
+                : undefined
+              }
```
Comment (johnleider) [inline packages/vuetify/src/components/VTimeline/VTimelineDivider.tsx:77]:
```suggestion
              {  slots.default
                 ? slots.default({ icon: props.icon, iconColor: props.iconColor, size: props.size })
                 : props.icon && (
                   <VIcon
                     color={ props.iconColor }
                     icon={ props.icon }
                     size={ props.size }
                   />
                 )
              }
```

File: packages/vuetify/src/components/VTimeline/VTimelineItem.tsx [hunk 5]
```diff
@@ -0,0 +1,99 @@
+// Types
+import type { ComponentPublicInstance } from 'vue'
+
+// Components
+import { VTimelineSymbol } from './VTimeline'
+import VTimelineDivider from './VTimelineDivider'
```
Comment (johnleider) [inline packages/vuetify/src/components/VTimeline/VTimelineItem.tsx:6]:
```suggestion
import { VTimelineDivider } from '.'
import { VTimelineSymbol } from './VTimeline'
```

File: packages/vuetify/src/components/VTimeline/VTimelineItem.tsx [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+
+// Composables
+import { makeTagProps } from '@/composables/tag'
+import { makeSizeProps } from '@/composables/size'
+import { makeElevationProps } from '@/composables/elevation'
+import { makeRoundedProps } from '@/composables/rounded'
+
+// Helpers
```
Comment (johnleider) [inline packages/vuetify/src/components/VTimeline/VTimelineItem.tsx:14]:
```suggestion
// Utilities
```

File: packages/vuetify/src/components/VTimeline/VTimelineItem.tsx [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+import { convertToUnit, makeProps } from '@/util'
+import { makeDimensionProps, useDimension } from '@/composables/dimensions'
+
+export default defineComponent({
+  name: 'VTimelineItem',
+
+  props: makeProps({
+    dotColor: {
+      type: String,
+      default: 'primary',
```
Comment (johnleider) [inline packages/vuetify/src/components/VTimeline/VTimelineItem.tsx:25]: remove default

File: packages/vuetify/src/components/VTimeline/VTimelineItem.tsx [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+    icon: String,
+    iconColor: String,
+    ...makeRoundedProps(),
+    ...makeElevationProps(),
+    ...makeSizeProps(),
+    ...makeTagProps(),
+    ...makeDimensionProps(),
+  }),
+
+  setup (props, ctx) {
```
Comment (johnleider) [inline packages/vuetify/src/components/VTimeline/VTimelineItem.tsx:42]:
```suggestion
  setup (props, { slots }) {
```

File: packages/vuetify/src/components/VTimeline/VTimelineItem.tsx [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+        style={{
+          // @ts-ignore
+          '--v-timeline-dot-size': convertToUnit(dotSize.value),
+        }}
+      >
+        <div
+          class="v-timeline-item__body"
+          style={ dimensionStyles.value }
+        >
+          { ctx.slots.default?.() }
```
Comment (johnleider) [inline packages/vuetify/src/components/VTimeline/VTimelineItem.tsx:75]:
```suggestion
          { slots.default?.() }
```

File: packages/vuetify/src/components/VTimeline/VTimelineItem.tsx [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+          elevation={ props.elevation }
+          dotColor={ props.dotColor }
+          fillDot={ props.fillDot }
+          rounded={ props.rounded }
+          v-slots={{ default: ctx.slots.icon }}
+        />
+
+        { timeline.density.value !== 'compact' && (
+          <div class="v-timeline-item__opposite">
+            { !props.hideOpposite && ctx.slots.opposite?.() }
```
Comment (johnleider) [inline packages/vuetify/src/components/VTimeline/VTimelineItem.tsx:93]:
```suggestion
            { !props.hideOpposite && slots.opposite?.() }
```

File: packages/vuetify/src/components/VTimeline/VTimelineItem.tsx [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+          ref={ dotRef }
+          hideDot={ props.hideDot }
+          icon={ props.icon }
+          iconColor={ props.iconColor }
+          size={ props.size }
+          elevation={ props.elevation }
+          dotColor={ props.dotColor }
+          fillDot={ props.fillDot }
+          rounded={ props.rounded }
+          v-slots={{ default: ctx.slots.icon }}
```
Comment (johnleider) [inline packages/vuetify/src/components/VTimeline/VTimelineItem.tsx:88]:
```suggestion
          v-slots={{ default: slots.icon }}
```

File: packages/vuetify/src/components/VTimeline/_variables.scss [hunk 12]
```diff
@@ -1,18 +1,29 @@
 @import '../../styles/styles.sass';
```
Comment (johnleider) [inline packages/vuetify/src/components/VTimeline/_variables.scss:2]:
```suggestion
// Defaults
```

**Review by johnleider — CHANGES_REQUESTED** [review]
* To avoid saying the same thing over and over, quite a bit of SASS needs to be updated to use the `&` selector.
* As far as styles, I'd like to remove the default white border and elevation from dots.
* re-structure sass to include an `_index.scss` file

File: packages/vuetify/src/components/VTimeline/VTimeline.tsx [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+      const side = props.side ? props.side : props.density !== 'default' ? 'after' : null
+
+      return side && `v-timeline--side-${side}`
+    })
+
+    return () => (
+      <props.tag
+        class={[
+          'v-timeline',
+          `v-timeline--${props.direction}`,
```
[the remark this answers is not in the document]
Comment (nekosaur) [reply packages/vuetify/src/components/VTimeline/VTimeline.tsx:83]: We don't do that any other place for template strings. Space between `{}` is for JSX syntax

File: packages/vuetify/src/components/VTimeline/VTimeline.tsx [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+    const { themeClasses } = useTheme()
+    const { densityClasses } = useDensity(props, 'v-timeline')
+
+    provide(VTimelineSymbol, {
+      density: toRef(props, 'density'),
+      lineColor: toRef(props, 'lineColor'),
+    })
+
+    const sideClass = computed(() => {
+      const side = props.side ? props.side : props.density !== 'default' ? 'after' : null
```
[the remark this answers is not in the document]
Comment (nekosaur) [reply packages/vuetify/src/components/VTimeline/VTimeline.tsx:74]: ?

image

**Comment by nekosaur** [comment]
> * As far as styles, I'd like to remove the default white border and elevation from dots.

remove as in remove the functionality altogether, or just have the default look be elevation 0 and fill-dot true?

File: packages/vuetify/src/components/VTimeline/_variables.scss [hunk 15]
```diff
@@ -1,18 +1,39 @@
...
+$timeline-density-comfortable-grid-template-end: min-content min-content auto !default;
+$timeline-density-comfortable-grid-template-start: auto min-content min-content !default;
+$timeline-density-compact-grid-template-end: min-content auto !default;
+$timeline-density-compact-grid-template-start: auto min-content !default;
+$timeline-divider-dot-elevation: 0 !default;
+$timeline-divider-line-background: rgba(var(--v-border-color), var(--v-border-opacity)) !default;
+$timeline-divider-line-horizontal-width: var(--v-timeline-line-inset, 0px) !default;
+$timeline-divider-line-thickness: var(--v-timeline-line-thickness) !default;
+$timeline-dot-border-radius: 50% !default;
+$timeline-dot-divider-background: rgb(var(--v-theme-on-surface)) !default;
```
Comment (nekosaur) [inline packages/vuetify/src/components/VTimeline/_variables.scss:11]: You can't use on color here, it won't work together with icons in the dot.

File: packages/vuetify/src/components/VTimeline/VTimelineDivider.tsx [hunk 16]
```diff
@@ -0,0 +1,85 @@
...
+
+export default defineComponent({
+  name: 'VTimelineDivider',
+
+  props: {
+    hideDot: Boolean,
+    lineColor: String,
+    icon: String,
+    iconColor: String,
+    fillDot: Boolean,
```
Comment (nekosaur) [inline packages/vuetify/src/components/VTimeline/VTimelineDivider.tsx:22]: You changed default look to be filled dot, but this prop still exist?

Comment (johnleider) [reply packages/vuetify/src/components/VTimeline/VTimelineDivider.tsx:22]: What does the default dot look like? Is it just white with the elevation?

File: packages/vuetify/src/components/VTimeline/_variables.scss (hunk 15 above)
Comment (johnleider) [reply packages/vuetify/src/components/VTimeline/_variables.scss:11]: Up for any suggestions. The dot should use surface color from theme.

Comment (nekosaur) [reply packages/vuetify/src/components/VTimeline/_variables.scss:11]: It can use the surface color. Not the surface on color.

Comment (nekosaur) [reply packages/vuetify/src/components/VTimeline/_variables.scss:11]: (also if it's supposed to be the surface color why can't default value for `dot-color` just be `surface`?)

Comment (johnleider) [reply packages/vuetify/src/components/VTimeline/_variables.scss:11]: As a matter of process with assigning default colors via prop. That in itself is a larger discussion.

**Comment by johnleider** [comment]
I'm having issues with modifying the line-inset value, which I assume is the amount that the divider line is distances from the dots?

image

**Review by johnleider — CHANGES_REQUESTED** [review]
Overall everything looks good. Needs to change out rtl/ltr with margin/padding-inline-start/end.

**Comment by johnleider** [comment]
Please resolve conflicts.

**Review by KaelWD — APPROVED** [review]

**Review by johnleider — APPROVED** [review]

**Pull request merged**